extensions/ql-vscode/src/data-serialization/json-validator.ts

```diff
@@ -0,0 +1,67 @@
+import Ajv, { ValidateFunction } from "ajv";
+import { JsonSchemaTypes } from "./json-schema-types";
+import * as fs from "fs";
+import { JSONSchema7 as JSONSchema } from "json-schema";
+
+export class JsonValidator {
+  ajv: Ajv;
+  schemaTypes: string[];
+  schemas: { [key: string]: JSONSchema };
+  validators: { [key: string]: ValidateFunction };
+
+  constructor() {
+    this.ajv = new Ajv();
+    this.schemaTypes = JsonSchemaTypes;
+    this.schemas = this.loadSchemas();
+    this.validators = this.loadValidators();
+  }
+
+  public validate(object: unknown, typeName: string) {
+    const validate = this.loadValidator(typeName);
+
+    const valid = validate(object);
+
+    if (!valid) {
+      throw new Error(
+        `Object does not match the "${typeName}" schema: ${this.ajv.errorsText(
+          validate.errors,
+        )}`,
+      );
+    }
+    return object;
+  }
+
+  private createValidatorFunction(typeName: string) {
+    const schema = this.loadSchema(typeName);
+    return this.ajv.compile(schema);
+  }
+
+  private loadSchemas(): { [key: string]: JSONSchema } {
+    const schemaList: { [key: string]: JSONSchema } = {};
+
+    for (const typeName of this.schemaTypes) {
+      const schemaFilepath = `src/data-serialization/generated-schemas/${typeName}.json`;
+      schemaList[typeName] = JSON.parse(
+        fs.readFileSync(schemaFilepath, "utf-8"),
```

Comment:
We probably want to avoid synchronous file reads in user-facing code. Maybe the schemas can be loaded asynchronously in a separate `init()` method, or as needed from `loadSchema`

Reply:
Thanks @aeisenberg! That's a good observation.

There's actually a follow up PR to this where we store the `JsonValidator` globally on the new `App` instance: https://github.com/github/vscode-codeql/pull/2020. 

This means we initialize the `JsonValidator` once in the extension (and perform all the schema file reading upfront) when the extension starts up, then re-use those schemas when we need to perform validation. 

So the file reading should only run once when the app is initialized rather than start reading files when a user performs a read/write action. 

Based on this, I think it might affect your original observation that we're reading files in user-facing code, but I might not be understanding all the sides to this so I'm interested to hear your interpretation.

If I move the file reading into an `init()` method, I'll still need to call it in the same place where I'm building the JsonValidator [when we initialize the app](https://github.com/github/vscode-codeql/pull/2020/commits/5fb787f3b19384c849dbfa818698c3d57d4057f8#diff-d75026577deda4d24fc20ee90b6db54add17aaefd2c2702a0d55a5fe04d6930bR13). 

`loadSchema()` is actually triggered when a user performs an action, so I think we would be performing file reading in user-facing code if we moved the file reading action in that method.

Reply:
I'm going to go ahead and merge this as-is, but if we discuss this further and decide it's still worth doing, I can open a a follow-up PR for it.

Reply:
Object constructors are not allowed to be `async`, so my suggestion was around moving the file reads outside of the constructor and put it into an `init` method so that it can be async. Startup time is already quite slow (especially if you have many databases in your workspace) and something we should probably look at some point. So, we should be careful about adding any more bottlenecks to startup.

Feel free to merge and I will look at your followup.